helper/cli.c

```diff
@@ -16,37 +16,162 @@
 #include <errno.h>
 #include <poll.h>
 #include <stdio.h>
+#include <strings.h>
 
 /* Socketpair socket roles. */
 enum {
 	SP_READ = 0,
 	SP_WRITE = 1,
 };
 
+#define MAX_NAME_LEN 20
+#define MAX_HELP_LEN 100
+
+typedef struct {
+	odph_cli_user_cmd_func_t fn;
+	char name[MAX_NAME_LEN];
+	char help[MAX_HELP_LEN];
+} user_cmd_t;
+
 typedef struct {
 	volatile int cli_fd;
 	/* Server thread will exit if this is false. */
 	volatile int run;
 	/* Socketpair descriptors. */
 	int sp[2];
 	int listen_fd;
-	/* This lock guards cli_fd and run, which must be accessed atomically. */
+	/* Guards cli_fd and run, which must be accessed atomically. */
 	odp_spinlock_t lock;
+	odp_spinlock_t api_lock;
 	odph_thread_t thr_server;
+	odp_instance_t instance;
+	struct sockaddr_in addr;
+	uint32_t max_user_commands;
+	uint32_t num_user_commands;
+	user_cmd_t user_cmd[0];
 } cli_shm_t;
 
 static const char *shm_name = "_odp_cli";
 
 static const odph_cli_param_t param_default = {
 	.address = "127.0.0.1",
 	.port = 55555,
+	.max_user_commands = 50,
 };
 
 void odph_cli_param_init(odph_cli_param_t *param)
 {
 	*param = param_default;
 }
 
+static cli_shm_t *shm_lookup(void)
+{
+	cli_shm_t *shm = NULL;
+	odp_shm_t shm_hdl = odp_shm_lookup(shm_name);
+
+	if (shm_hdl != ODP_SHM_INVALID)
+		shm = (cli_shm_t *)odp_shm_addr(shm_hdl);
+
+	return shm;
+}
+
+int odph_cli_init(odp_instance_t instance, const odph_cli_param_t *param)
+{
+	if (odp_shm_lookup(shm_name) != ODP_SHM_INVALID) {
+		ODPH_ERR("Error: shm %s already exists\n", shm_name);
+		return -1;
+	}
+
+	cli_shm_t *shm = NULL;
+	int shm_size = sizeof(cli_shm_t) +
+		param->max_user_commands * sizeof(user_cmd_t);
+	odp_shm_t shm_hdl =
+		odp_shm_reserve(shm_name, shm_size, 64, ODP_SHM_SW_ONLY);
+
+	if (shm_hdl != ODP_SHM_INVALID)
+		shm = (cli_shm_t *)odp_shm_addr(shm_hdl);
+
+	if (!shm) {
+		ODPH_ERR("Error: failed to reserve shm %s\n", shm_name);
+		return -1;
+	}
+
+	memset(shm, 0, shm_size);
+	odp_spinlock_init(&shm->lock);
+	odp_spinlock_init(&shm->api_lock);
+	shm->sp[SP_READ] = -1;
+	shm->sp[SP_WRITE] = -1;
+	shm->listen_fd = -1;
+	shm->cli_fd = -1;
+	shm->instance = instance;
+
+	shm->addr.sin_family = AF_INET;
+	shm->addr.sin_port = htons(param->port);
+
+	switch (inet_pton(AF_INET, param->address, &shm->addr.sin_addr)) {
+	case -1:
+		ODPH_ERR("Error: inet_pton(): %s\n", strerror(errno));
+		return -1;
+	case 0:
+		ODPH_ERR("Error: inet_pton(): illegal address format\n");
+		return -1;
+	}
```

Comment:
Some compilers or static code checkers may complain about a switch statement that does not have a default case. So, I'd suggest to add default, or change this into two if-clauses.

Reply:
Added default case in v7.

Reply:
Checkpatch requires break also:

```
default:
        break;
```

Reply:
Fixed in v8.